Certificate revision history limit

JoshVanL opened the pull request:
fixes  #3458
/assign @jakexks 
/kind feature

This PR adds a new Certificate controller that is responsible for cleaning up old CertificateRequests using a new `RevisionHistoryLimit` field.

This optional field will ensure that only x number of old request revisions for a Certificate exist. The field is set to `nil` by default, which won't clean up any CRs. The controller will delete CRs from oldest first, and will ignore CRs if they don't have the annotation or have an invalid value (which shouldn't normally happen). The controller won't trigger if the Certificate isn't in a Ready state so as to not conflict with other processing. The value must be set to 1 or greater- this is to ensure we keep _some_ kind of visibility and doesn't cause potential issues for other controllers.

I'm not 100% about the "revisionmanager' name, so keen to hear suggestions :smile: 


```release-note
Adds RevisionHistoryLimit field to Certificates to optionally garbage collect old CertificateRequests
```

JoshVanL commented:
/retest

JoshVanL commented:
/assign @RiichardCC

JoshVanL commented:
Whoops, sorry!

/unassign @RiichardCC 
/assign @wallrj

wallrj commented on test/integration/certificates/revisionmanager_controller_test.go:
```diff
@@ -0,0 +1,184 @@
+		mustSync,
+		nil,
+		queue,
+	)
```
That's cool. I'd never investigated the integration tests before. Nice to be able to run a single controller against a real API server.

wallrj commented on test/integration/certificates/revisionmanager_controller_test.go:
```diff
@@ -0,0 +1,184 @@
+
+	c := controllerpkg.NewController(
+		context.Background(),
+		"issuing_test",
```
```suggestion
		"revisionmanager_controller_test",
```

wallrj reviewed (changes requested):
Thanks @JoshVanL 

It looks like great code and tests, but I am not convinced that the revision history should be configurable per-certificate, and I'm personally in favour of deleting all non-latest revision CRs, which seems to be what the old-style controllers did.
What do you think?

wallrj commented:
/unassign

JoshVanL commented on pkg/controller/certificates/revisionmanager/revisionmanager_controller.go:
```diff
@@ -0,0 +1,216 @@
+
+	limit := int(*crt.Spec.RevisionHistoryLimit)
+
+	// Only garbage collect over Certificates that are in a Ready=True condition.
+	if !apiutil.CertificateHasCondition(crt, cmapi.CertificateCondition{
+		Type:   cmapi.CertificateConditionReady,
+		Status: cmmeta.ConditionTrue,
+	}) {
+		return nil
+	}
```
In the case where we have a string of Failed CRs and the Cert is still Failed, it would be prudent to expect the user would want to check the CR history and their failure reasons.

JoshVanL commented on pkg/controller/certificates/revisionmanager/revisionmanager_controller.go:
```diff
@@ -0,0 +1,216 @@
+		log = logf.WithRelatedResource(log, req)
+
+		if req.Annotations == nil || req.Annotations[cmapi.CertificateRequestRevisionAnnotationKey] == "" {
+			log.V(logf.DebugLevel).Info("skipping processing request with missing revsion")
```
I think it makes sense that if user changes this, they might want it to be around for whatever reason. I believe we should be as permissive as possible here.

JoshVanL commented on pkg/controller/certificates/revisionmanager/revisionmanager_controller.go:
```diff
@@ -0,0 +1,216 @@
+
+	sort.SliceStable(revisions, func(i, j int) bool {
+		return revisions[i].rev < revisions[j].rev
+	})
```
Yes, but we have to re-parse the int every time, which would be O(n{something :sweat_smile: })

JoshVanL commented on pkg/controller/certificates/revisionmanager/revisionmanager_controller_test.go:
```diff
@@ -0,0 +1,386 @@
+				testpkg.NewAction(coretesting.NewDeleteAction(cmapi.SchemeGroupVersion.WithResource("certificaterequests"), "testns", "cr-1")),
+				testpkg.NewAction(coretesting.NewDeleteAction(cmapi.SchemeGroupVersion.WithResource("certificaterequests"), "testns", "cr-2")),
+				testpkg.NewAction(coretesting.NewDeleteAction(cmapi.SchemeGroupVersion.WithResource("certificaterequests"), "testns", "cr-6")),
+			},
```
Perhaps, though that is a lot more syncs that will be called as a result. I prefer doing it all in one sync.

JoshVanL commented:
@wallrj Hopefully have addressed your comments. As @munnerz has suggested, I have left this default to nil to keep the same behaviour of the v1.0 release.

wallrj reviewed (approved):
Thanks for the answers and changes @JoshVanL 

I still think the default should be 1 and from his comment  maybe @munnerz thinks the same.
i.e. existing certificates should default to nil. New certificates should default to 1.
Then we can legitimately say that we fixed #3458 in which @meyskens  wrote:

> Expected behaviour:
> The controller deleting unused CertificateRequests after a while

Personally I don't buy the v1 compatibility argument.
We break this rule when it suits us, for example by adding validation to make CRs immutable.
What about the hypothetical user who relies on being able to modify the spec of a failed CR that they created directly?
If we followed the rule strictly we'd hardly ever fix any bugs in cert-manager.

All we guarantee is that cert-manager will operate on a `Certificate` and populate a `Secret`; how it achieves that should be an implementation detail.

And that this bug-fix / work-around introduces a new public API field also seems fishy.
If we agree that the long-term goal is to have cert-manager delete all but the latest CR for a Certificate, then we could instead log a warning and an event when multiple CRs are detected, with a message that directs users to a documentation page describing a `--feature-delete-stale-certificate-requests` flag which will become the default in a couple of releases.

Having said all that, 

/lgtm
/hold
(in case you want to reconsider)

wallrj reviewed (approved):
/lgtm

JoshVanL commented:
/hold cancel
